packages/@react-aria/menu/src/useSafelyMouseToSubmenu.ts

```diff
@@ -0,0 +1,94 @@
+import {CSSProperties, RefObject, useEffect, useRef, useState} from 'react';
+import {useResizeObserver} from '@react-aria/utils';
+
+interface SafelyMouseToSubmenuOptions {
+  /** Ref for the parent menu. */
+  menuRef: RefObject<Element>,
+  /** Ref for the submenu. */
+  submenuRef: RefObject<Element>,
+  /** Whether the submenu is open. */
+  isOpen: boolean
+}
+
+/**
+ * Allows the user to move their pointer to the submenu without it closing when their mouse leaves the trigger element.
+ * Prevents pointer events from going to the underlying menu if the user is moving their pointer towards the sub-menu.
+ */
+export function useSafelyMouseToSubmenu(options: SafelyMouseToSubmenuOptions): CSSProperties {
+  let {menuRef, submenuRef, isOpen} = options;
+  let prevPointerPos = useRef<{x: number, y: number}>(null);
+  let submenuRect = useRef<DOMRect>(null);
+  let lastProcessedTime = useRef<number>(0);
+  let [isPointerMovingTowardsSubmenu, setIsPointerMovingTowardsSubmenu] = useState(false);
+
+  let updateSubmenuRect = () => {
+    if (submenuRef.current) {
+      submenuRect.current = submenuRef.current.getBoundingClientRect();
+    }
+  };
+
+  useResizeObserver({ref: submenuRef, onResize: updateSubmenuRect});
+
+  useEffect(() => {
+    let menu = menuRef.current;
+    let submenu = submenuRef.current;
+
+    if (!menu || !submenu || !isOpen) {
+      return;
+    }
+
+    submenuRect.current = submenu.getBoundingClientRect();
+
+    let onPointerMove = (e: PointerEvent) => {
+      let currentTime = Date.now();
+
+      // Throttle
+      if (currentTime - lastProcessedTime.current < 16) {
+        return;
+      }
+
+      let prevIsPointerMovingTowardsSubmenu = isPointerMovingTowardsSubmenu;
+
+      let {clientX: mouseX, clientY: mouseY} = e;
+
+      if (!prevPointerPos.current) {
+        prevPointerPos.current = {x: mouseX, y: mouseY};
+        return;
+      }
+
+      if (!submenuRect.current) {
+        return;
+      }
```

Comment:
Discussed this in DMs, but could try setting `isPointerMovingTowardsSubmenu` to false here to try and get rid of the weird race condition we are seeing